File: src/main/java/io/github/cdgeass/editor/provider/XmlLineMarkerProvider.java
package io.github.cdgeass.editor.provider;

import com.intellij.codeInsight.daemon.RelatedItemLineMarkerInfo;
import com.intellij.codeInsight.daemon.RelatedItemLineMarkerProvider;
import com.intellij.codeInsight.navigation.NavigationGutterIconBuilder;
import com.intellij.icons.AllIcons;
import com.intellij.psi.PsiElement;
import com.intellij.psi.xml.XmlFile;
import io.github.cdgeass.editor.dom.DomUtil;
import io.github.cdgeass.editor.dom.element.mapper.Mapper;
import io.github.cdgeass.editor.dom.element.mapper.Statement;
import org.jetbrains.annotations.NotNull;

import java.util.Collection;
import java.util.Collections;
import java.util.Optional;
import java.util.function.Consumer;

/**
 * @author cdgeass
 * @since 2020-05-17
 */
public class XmlLineMarkerProvider extends RelatedItemLineMarkerProvider {

    @Override
    protected void collectNavigationMarkers(@NotNull PsiElement element, @NotNull Collection<? super RelatedItemLineMarkerInfo<?>> result) {
        if (!(element instanceof XmlFile)) {
            return;
        }

        var mapper = DomUtil.findDomElement((XmlFile) element, Mapper.class);
        if (mapper == null || mapper.getXmlTag() == null) {
            return;
        }
        var namespaceAttributeValue = mapper.getNamespace();
        var psiClass = namespaceAttributeValue.getValue();
        if (psiClass == null || psiClass.getQualifiedName() == null) {
            return;
        }
        var rootIconBuilder = NavigationGutterIconBuilder.create(AllIcons.Gutter.ImplementingMethod)
                .setTarget(psiClass)
                .setTooltipText(psiClass.getQualifiedName());
        result.add(rootIconBuilder.createLineMarkerInfo(mapper.getXmlTag().getNavigationElement()));

        Consumer<Statement> consumer = statement -> {
            if (statement.getXmlTag() == null) {
                return;
            }
            var methodAttributeValue = statement.getId();
            var psiMethod = methodAttributeValue.getValue();
            if (psiMethod == null) {
                return;
            }
            NavigationGutterIconBuilder<PsiElement> subIconBuilder = NavigationGutterIconBuilder.create(AllIcons.Gutter.ImplementingMethod)
                    .setTarget(psiMethod)
                    .setTooltipText(psiMethod.getName());
            result.add(subIconBuilder.createLineMarkerInfo(statement.getXmlTag().getFirstChild()));
        };
        Optional.ofNullable(mapper.getSelects()).orElse(Collections.emptyList()).forEach(consumer);
        Optional.ofNullable(mapper.getDeletes()).orElse(Collections.emptyList()).forEach(consumer);
        Optional.ofNullable(mapper.getInserts()).orElse(Collections.emptyList()).forEach(consumer);
        Optional.ofNullable(mapper.getUpdates()).orElse(Collections.emptyList()).forEach(consumer);
    }
}

File: src/main/kotlin/io/github/cdgeass/codeInsight/daemon/XmlLineMarkerProvider.kt
package io.github.cdgeass.codeInsight.daemon

import com.intellij.codeInsight.daemon.RelatedItemLineMarkerInfo
import com.intellij.codeInsight.daemon.RelatedItemLineMarkerProvider
import com.intellij.codeInsight.navigation.NavigationGutterIconBuilder
import com.intellij.icons.AllIcons
import com.intellij.psi.PsiClass
import com.intellij.psi.PsiElement
import com.intellij.psi.util.PsiTreeUtil
import com.intellij.psi.util.elementType
import com.intellij.psi.xml.XmlTag
import com.intellij.psi.xml.XmlTokenType
import com.intellij.util.xml.DomUtil
import io.github.cdgeass.editor.dom.element.mapper.Mapper
import io.github.cdgeass.editor.dom.element.mapper.Statement

/**
 * @author cdgeass
 * @since 2021-03-18
 */
class XmlLineMarkerProvider : RelatedItemLineMarkerProvider() {

    override fun getName(): String {
        return "MYBATIS XML"
    }

    override fun collectNavigationMarkers(
        element: PsiElement,
        result: MutableCollection<in RelatedItemLineMarkerInfo<*>>
    ) {
        if (element.elementType != XmlTokenType.XML_NAME) {
            return
        }

        val lineMarkerInfo: RelatedItemLineMarkerInfo<PsiElement> = when {
            isRootToken(element) -> {
                collectRootTarget(element)
            }
            isStatementToken(element) -> {
                collectTarget(element)
            }
            else -> {
                null
            }
        } ?: return

        result.add(lineMarkerInfo)
    }

    private fun collectRootTarget(element: PsiElement): RelatedItemLineMarkerInfo<PsiElement>? {
        val mapper = DomUtil.findDomElement(element, Mapper::class.java) ?: return null

        val psiClass = mapper.namespace.value ?: return null
        return NavigationGutterIconBuilder.create(AllIcons.Gutter.ImplementingMethod)
            .setTarget(psiClass.identifyingElement)
            .createLineMarkerInfo(element)
    }

    private fun collectTarget(element: PsiElement): RelatedItemLineMarkerInfo<PsiElement>? {
        val statement = DomUtil.findDomElement(element, Statement::class.java) ?: return null

        val psiMethod = statement.id.value ?: return null
        val psiClass = PsiTreeUtil.findFirstParent(psiMethod) { psiElement ->
            psiElement is PsiClass
        } ?: return null

        // 如果 xml 指向的是父类方法 直接定位子类上 TODO 增加可选项
        val mapper = DomUtil.findDomElement(element, Mapper::class.java) ?: return null
        val identifier = if (psiClass == mapper.namespace.value) {
            psiMethod.identifyingElement
        } else {
            mapper.namespace.value?.identifyingElement
        } ?: return null

        return NavigationGutterIconBuilder.create(AllIcons.Gutter.ImplementingMethod)
            .setTarget(identifier)
            .createLineMarkerInfo(element)
    }

    private fun isRootToken(element: PsiElement): Boolean {
        return element.prevSibling.elementType == XmlTokenType.XML_START_TAG_START && element.text == "mapper"
    }

    private fun isStatementToken(element: PsiElement): Boolean {
        return element.text == "id" && PsiTreeUtil.findFirstParent(element) {
            it is XmlTag && STATEMENT_NAMES.contains(it.name)
        } != null
    }

    companion object {
        private val STATEMENT_NAMES = listOf("select", "update", "delete", "insert")
    }
}